### DISCUSSION TITLE
Suggestion: Database UI

### DISCUSSION THREAD

**Discussion by osseonews** [nhost/nhost#965] [Ideas]
Love the new database UI. Only one issue: If a column is a foreign key, the UI actually puts the primary key value from the relationship table in the UI, and if you want to edit the Foreign Key value you then need to find the actual primary key value in the related Table. So let's say we have a Posts Table and an Author Column in the Posts Table which is a foreign key to the Author Table, the Posts table will show an integer in the UI as a foreign key. While this is accurate on the database level, for a UI it is very, very difficult to use. What should happen is that the UI should display a chosen field from the Author table that corresponds to the key in the actual database. It the example, you will see the Author Name, instead of the actual foreign key value to the Author table. That way you can tell via the UI what is actually going on the database. As well, when you create a foreign key via the UI, you shouldn't have to figure out what the primary key value is for the foreign key, instead you should be given a list of choices from the foreign table that is readable, like Author Name. Then after choosing the UI field, the underlying code can assign the correct key.

Also, one other thing: Any chance you can open source the Database UI, so we can use it for staff members that are not logging into  the Nhost Console? The idea here would be that they can just edit rows, but are not able to add/edit tables/columns etc.

**Comment by szilarddoro** [comment]
Hey @osseonews, we're glad that you like using our new Database UI.

**About the Foreign Keys and their visual representation:**
Thanks for raising your concern, the Author - Posts relationship a very good example. This matter was actually on the table once, but we decided to postpone it for a while. Unfortunately it's not trivial which field to pick when displaying the "referenced" item instead of the foreign key. We don't have an additional layer above PostgreSQL yet which could be used to let users decide which field to use as a "label" for that foreign key.

What I can think of here is either to put a "View" button next to the referenced item's identifier or to turn the identifier into an anchor. Both of these would then redirect the user to the referenced item.

As for the row creator / inline editing, we see the value of showing available items in a dropdown instead of "forcing" users to search for the correct ID in the referenced table. But again, deciding which field to pick as a "label" is not trivial, so we need to come up with some solution for this.

**About open sourcing the Database UI:**
You can read a few words about this in [our blog post](https://nhost.io/blog/new-database-ui). We are working hard on open sourcing the dashboard soon, but we are not there yet. Unfortunately, we don't have ETA on this, but will keep everyone posted.

**Reply by osseonews** [reply]
Thanks for the quick reply. I know how difficult a good UI is, which is why I think this is essential for any database and this is the main thing Hasura is missing. The solution for the Foreign Key representation is to just keep it simple at the start so the UI is actually usable for foreign keys. It can be improved over time b/c as of right now the UI is not really functional when it comes to foreign keys. The simplest solution is as follows:
1. Click on a foreign key field, should be bring up a View in a modal, which shows the data from the foreign key record. You definately don't want to redirect the user, as the user is viewing the current table, and doesn't expect to redirect to another table. They just want see the underlying data that the foreign key represents. The modal for viewing a foreign key record, can look exactly like the "Insert New Record" modal you create when a user wants to add a new Record. You can even allow this modal to open an Edit View of the entire record, in case a user wants to edit the foreign key data also (or have a toggle on the view to allow Edit). So you have the correct Modal ui now, it's just not used for foreign keys.
2.  Create/Edit Foreign Key: Just use a "name" or "title" or "label" field as the reference field for a Label and provide instruction that in order to create/edit a foreign key, you need a "required" name or title in the foreign key table, or simply require this field for every table when creating a table via the UI. Yeah, this forces people to add a name or title field to a table, but big deal. The improvement in the UI is worth it. Over time, you can build in code to allow the user to choose which field to act as a Label, but to start just fix the field for usability concerns.

BTW, three other UI correction
1.  You should immediately add created_at and updated_at fields for every new table created via the UI, without having to force a user to do this manually. This will save a ton of time for anyone building tables.
2. You should also require a "name" or "title" field to use as Label for foreign keys. See above. Or simply call this field "label"
3.  If you have alot of columns the table view will get overwhelming, so you need some sort of UI tool to allow the user to choose which columns are visible in the UI. 

Finally, a suggestion. Check out the amazing React-Admin library for inspiration. They have solved most of the problems related to UI's for dashboards.

**Reply by szilarddoro** [reply]
**Foreign record opened in a modal:**

Yes, I think this is what we want to go with in the beginning, then reiterate later.

**Using `name` / `title` / `label` field as the "identifier" field:**

I'm not sure about this. I would definitely not force users to follow our guidelines and make this a mandatory field in every table. We should add some custom implementation on our end to let the user decide which field to use as an "identifier" from the existing ones.

**Adding `created_at` and `updated_at` columns by default:**

I don't think it's a good idea to add these columns (or any column in general) to every table by default (e.g: doesn't make too much sense for junction tables).

What we should do instead is to let users decide when they want to add them and make it as easy as possible. We already have a feature similar to Hasura's "Frequently Used Columns" on our roadmap. This would speed up the manual (and very slow) process of column creation.

**Reply by osseonews** [reply]
"I would definitely not force users to follow our guidelines". Agree you don't want to force users. What you do is that if there is a "label" field in the foreign table, you use that as the "identifier" field, if not, then you use the primary key. So users then have a choice. Of course, there is a more complicated way to do this longer-term, but for start, I think an "optional" label field in a table would greatly improve the UI for foreign keys.

**Reply by pkreipke** [reply]
First off, the new DB UX is great, love it as an alternative for simple edits of our data. However, the Hasura interface is fine for the missing functionality IMO and with direct DB access anyone can stand up their own 'dashboard'. I would personally rather see other new features get attention rather than miss out on opportunities because resources were assigned to make a better DB 'dashboard'.

That said, I don't know the NHost business plan or priorities [hint - share them ;-) ] so comments on the above.

Open in Modal: I agree, that'd be a nice feature. But I'd be happy to have simple nav open up the target in a new tab instead. I'll figure that out as a user :-).

'label' column: I agree with @szilarddoro - I wouldn't want to pay this price on every table, nor on every row of that table. In addition, I wouldn't want the field name to be out of use to me - so at the very least it should be something obscure like '_xyz' or 'ux_label_value'. Either way, I feel like it should be a table level attribute not in the Postgres schema. Or perhaps from a comment on the field? If it could be hoisted to the Hasura level as an attribute of the table's metadata, that might be ok. Maybe hint at that over in Hasura's Discord as a new feature request.

'created_at' / 'updated_at': also agree. Hasura's helper is nice. Not requiring them is great.

My $0.02.

**Reply by osseonews** [reply]
Agree somewhat with what you said, but in terms of this comment "IMO and with direct DB access anyone can stand up their own 'dashboard'." 

Yeah, anyone can develop their own dashboard. But is it worth the time/effort to build it and maintain it? Heck No. Nobody likes building admin dashboards. They want to focus on setting up the tables and working on the business logic. This is why there are so many companies offering this very service as a Saas, and why no-code companies like Airtable are growing like crazy.  Personally, I think the core Hasura UI is garbage and Hasura has never improved it at all, so I am very happy Nhost is investing in a better UI, and I hope they just keep improving it based on suggestions.

I think that the first company that offers a good database UI tool out of the box with the database service will grow massively in the coming years, b/c this is the huge, huge void in the market now. If you want a good UI, then you inevitably move to a nocode tool, like Airtable, but then you lose all benefits of a good database, so it's a terrible decision. If want a real database, well then you need to build your own dashboard UI, which is pure hell long-term, especially since all these dashboards are the same . Nhost made a tremendously smart decision to develop on Hasura, so alot of the functionality their competitors are building is already available in core Hasura out of the box. This allows them to focus on adding more value-add features, as opposed to basic features that are somewhat difficult on other platforms.

**Reply by pkreipke** [reply]
I chose my words with care: "stand up" a dashboard not develop. I agree with you - would never go build over buy for a DB editor like this. 

Again, IMO only, this is far from the most important missing feature. More detailed, automated logging and export to external systems like Datadog etc are far more important for a complete BaaS like NHost for example.

Also, I will likely never need a direct DB editor for _internal users_ and I suspect that's true even for your use cases. As a dev or admin the DB views/edits you describe could be made either through Hasura for the more advanced ones, or by writing code (GQL or PG or functions) [ hear me out ;-) ].

Any view/edits that would be done by non-devs would require knowledge of the business rules that are needed to make intelligent queries or edits. E.g. I'd likely make DB access available through a proper dashboard like React-Admin or ForestAdmin etc.

End of pitch for other priorities :-)

**Reply by osseonews** [reply]
Really appreciate this discussion. You are right in certain respects and actually for internal users, we will likely have to built a dashboard for our current use case, and I'm dreading this frankly, even if I really like React-Admin. However, what I'm hoping for is that the regular admin UI can at least get us to launch and then I'll worry about the internal UI after. 

So maybe my pitch is just for them to open source this UI. Maybe this would make it easier than starting the dashboard on our own.

**Reply by pkreipke** [reply]
👍 

Another thing to remember (Nhost): whether the implementation is tied to Hasura or direct to PG DB, it'll still need maintenance and sync with upstream Hasura/PG schema changes. E.g. more resources.

**Reply by elitan** [reply]
Thanks for your thoughtful answers in this exciting discussion. 

I would agree that it would be a bad decision to build a Database UI since we already provide one via the Hasura Console. And that's what we've been doing since the start (2019) of Nhost. 

However, we've reached a point where we think it does make sense to build the database UI in-house. Here are a few reasons:

- **Fewer Concepts:** Having two systems (Nhost Dashboard and Hasura Console) can be confusing for first-time users and creates a lot of friction in getting started. It's better to have one (Nhost Dashboard) as it's fewer concepts to learn and less context switching.
- **Competitors**: Hasura Console is pretty good, but not as good as some of our competitor's database UI's (Supabase, PocketBase, Airtable, Baserow, etc.). Developers spend a lot of time in the database UI developing their apps, and it's essential for us that the experience is world-class.
- **Innovation:** With the Nhost Database UI, we can innovate in a way that was not previously possible.

> Again, IMO only, this is far from the most important missing feature. More detailed, automated logging and export to external systems like Datadog etc are far more important for a complete BaaS like NHost for example.

Valid point. The launch week is not done yet, and we might have something cooking in this direction later this week 👀.

> Another thing to remember (Nhost): whether the implementation is tied to Hasura or direct to PG DB, it'll still need maintenance and sync with upstream Hasura/PG schema changes. E.g. more resources.

True, and we're diligent and rather conservative in our implementation. E.g., I don't think the `CREATE TABLE` syntax will change anytime soon. 

Again, thanks for your perspectives. They are valuable to us.

**Reply by osseonews** [reply]
@elitan You mention PocketBase above. I've experimented alot with Pocketbase. It is excellent, while also being simple. Your UI is pretty similar, except for foreign key issue. Please just see what Pocketbase handles the Foreign Key issue (they call it a Relation field) in the UI and if you implemented that way, I think it would be more than sufficient for now.  Thanks.

**Comment by osseonews** [comment]
BTW, one other small UI suggestion. For JSONB fields you can add in a JSON editor. Right now, we are copying the JSON field data into an online JSON editor to edit. Perhaps this package can work? https://github.com/json-editor/json-editor
